### PR TITLE
compute_ctl: Add endpoint that allows extensions to be installed

### PR DESCRIPTION
**Pull request by devjv**
Adds endpoint to install extensions:

**POST** `/extensions`
```
{"extension":"pg_sessions_jwt","database":"neondb","version":"1.0.0"}
```

Will be used by `local-proxy`.
Example, for the JWT authentication to work the database needs to have the pg_session_jwt extension and also to enable JWT to work in RLS policies.

### CODE REVIEW

File: compute_tools/src/http/api.rs [hunk 1]
```diff
@@ -98,6 +98,32 @@ async fn routes(req: Request<Body>, compute: &Arc<ComputeNode>) -> Response<Body
             }
         }
 
+        (&Method::POST, "/extensions") => {
+            info!("serving /extensions POST request");
+            let status = compute.get_status();
+            if status != ComputeStatus::Running {
```
Comment (devjv) [inline compute_tools/src/http/api.rs:104]: Not 100% sure if this is needed.

File: compute_tools/src/compute.rs [hunk 2]
```diff
@@ -1367,6 +1368,26 @@ LIMIT 100",
...
+    pub fn install_extension(&self, ext_name: &str, db_name: &str) -> Result<()> {
+        let mut conf = Config::from_str(self.connstr.as_str())
+            .context("Failed to parse connection string")?;
+        conf.dbname(db_name);
+    
+        let mut db_client = conf.connect(NoTls)
+            .context("Failed to connect to the database")?;
+    
+        let query = format!(
+            "CREATE EXTENSION IF NOT EXISTS {}",
```
Comment (devjv) [inline compute_tools/src/compute.rs:1380]: This would only work if the extension is available

File: compute_tools/src/http/api.rs [hunk 3]
```diff
@@ -98,6 +98,32 @@ async fn routes(req: Request<Body>, compute: &Arc<ComputeNode>) -> Response<Body
...
+                );
+                error!(msg);
+                return Response::new(Body::from(msg));
+            }
+
+            let body = hyper::body::to_bytes(req.into_body()).await.unwrap();
+            let body = serde_json::from_slice::<serde_json::Value>(&body).unwrap();
+            let extension = body["extension"].as_str().unwrap();
+            let database = body["database"].as_str().unwrap();
+            let res = compute.install_extension(extension, database);
```
Comment (devjv) [inline compute_tools/src/http/api.rs:117]: Does this need to be async?

File: compute_tools/src/http/api.rs [hunk 4]
```diff
@@ -98,6 +98,32 @@ async fn routes(req: Request<Body>, compute: &Arc<ComputeNode>) -> Response<Body
...
+            match res {
+                Ok(_) => Response::new(Body::from("true")),
+                Err(e) => {
+                    error!("install_extension failed: {}", e);
+                    Response::new(Body::from(e.to_string()))
```
Comment (conradludgate) [inline compute_tools/src/http/api.rs:122]: can we add a status code?

File: compute_tools/src/compute.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (devjv) [reply compute_tools/src/compute.rs:1381]: Should this implicitly upgrade the extension if called another time or only if the version is provided in the request?

[the remark this answers is not in the document]
Comment (devjv) [reply compute_tools/src/compute.rs:1381]: Also should we check if the extension is installed and return early or just have this simple for now and maybe add support for managing versions in further work cc @conradludgate

**Comment by hlinnaka** [comment]
What's the backstory behind this, why do we need this endpoint?

**Comment by devjv** [comment]
@hlinnaka 
> What's the backstory behind this, why do we need this endpoint?

For the JWT authentication to work the database needs to have the `pg_session_jwt` extension and also to enable JWT to work in RLS policies. Initial approach was to enable the extension as part of apply_config but an HTTP endpoint to do this was proposed by maintainers as a preferred solution.

File: compute_tools/src/compute.rs [hunk 5]
```diff
@@ -1367,6 +1368,38 @@ LIMIT 100",
...
+            .context("Failed to connect to the database")?;
+
+        let query = format!(
+            "CREATE EXTENSION IF NOT EXISTS {}",
+            ext_name.to_string().pg_quote()
+        );
+        info!("creating extension with query: {}", query);
+
+        db_client
+            .execute(&query, &[])
```
Comment (tristan957) [inline compute_tools/src/compute.rs:1387]: There is a `simple_query` method that should be easier to use.

File: compute_tools/src/compute.rs [hunk 6]
```diff
@@ -1367,6 +1368,38 @@ LIMIT 100",
...
+            .context(format!("Failed to execute query: {}", query))?;
+
+        let version_query = format!(
+            "SELECT extversion FROM pg_extension WHERE extname = '{}'",
+            ext_name.to_string().pg_quote()
+        );
```
Comment (tristan957) [inline compute_tools/src/compute.rs:1393]: Let's use a paramaterized query just to be safe.

File: compute_tools/src/compute.rs [hunk 7]
```diff
@@ -1367,6 +1368,38 @@ LIMIT 100",
         download_size
     }
 
+    pub fn install_extension(&self, ext_name: &str, db_name: &str) -> Result<String> {
+        let mut conf =
+            Config::from_str(self.connstr.as_str()).context("Failed to parse connection string")?;
```
Comment (tristan957) [inline compute_tools/src/compute.rs:1373]: I think it should be good to just `unwrap` here. At this point, we've used the same connection string many times to connect.

File: compute_tools/src/http/openapi_spec.yaml [hunk 8]
```diff
@@ -144,6 +144,45 @@ paths:
...
+          application/json:
+            schema:
+              type: object
+              required:
+                - extension
+                - database
+              properties:
+                extension:
+                  type: string
+                  description: Extension name.
```
Comment (tristan957) [inline compute_tools/src/http/openapi_spec.yaml:167]: Please also provide an example, even though it's trivial.

File: compute_tools/src/http/api.rs [hunk 9]
```diff
@@ -98,6 +99,38 @@ async fn routes(req: Request<Body>, compute: &Arc<ComputeNode>) -> Response<Body
             }
         }
 
+        (&Method::POST, "/extensions/install") => {
```
Comment (tristan957) [inline compute_tools/src/http/api.rs:102]: I think a `/extensions` endpoint would be better, but fairly subjective critique.

```
GET /extensions # list of all extensions. right now we are using /installed_extensions, but we can change that in the future
DELETE /extensions # uninstall an extension
POST /extensions # install an extension
```

File: compute_tools/src/http/api.rs [hunk 10]
```diff
@@ -98,6 +99,38 @@ async fn routes(req: Request<Body>, compute: &Arc<ComputeNode>) -> Response<Body
...
+                    status
+                );
+                error!(msg);
+                return Response::new(Body::from(msg));
+            }
+
+            let body = hyper::body::to_bytes(req.into_body()).await.unwrap();
+            let body = serde_json::from_slice::<serde_json::Value>(&body).unwrap();
+            let extension = body["extension"].as_str().unwrap();
+            let database = body["database"].as_str().unwrap();
```
Comment (tristan957) [inline compute_tools/src/http/api.rs:117]: Maybe we should have an `ExtensionInstallRequest` type?

**Comment by tristan957** [comment]
Thanks for doing this!

**Comment by devjv** [comment]
@tristan957 @MMeent thanks for the thorough review! I agree with your comments, will ping when changes are done. 👍

File: compute_tools/src/http/api.rs (hunk 9 above)
Comment (tristan957) [reply compute_tools/src/http/api.rs:102]: > Disagree with /extensions as name, as that implies plural, while this endpoint only handles a singular extension.

That is typical REST API design. See our public API for creating a project: https://api-docs.neon.tech/reference/createproject.

Comment (tristan957) [reply compute_tools/src/http/api.rs:102]: It also wouldn't be inconceivable to just take an array for the endpoint anyway.

Comment (devjv) [reply compute_tools/src/http/api.rs:102]: To me the most REST way would be just `POST` to `/extensions`, it implies we're adding/installing one. But also this is not a public API so it doesn't matter as much. I'm fine with either.

Comment (tristan957) [reply compute_tools/src/http/api.rs:102]: I've been wanting compute_ctl to grow a more REST-like interface anyway. This is a good start. I'd eventually like to take a dependency on something like axum to structure this code better.

Comment (tristan957) [reply compute_tools/src/http/api.rs:102]: > IMO, doing this at the top level of /extensions doesn't make much sense - we can't list the installed extensions without parameterizing on database or making the endpoint slow by O(n databases). A not-quite-REST endpoint under /extensions/install makes more sense to me.

Isn't this was Anastasia's work was all about? We just move here code into the `GET` side of the endpoint.

File: compute_tools/src/compute.rs (hunk 6 above)
Comment (devjv) [reply compute_tools/src/compute.rs:1393]: I assume you mean for all of the queries? `simple_query` doesn't appear to allow that? Many other queries in the same file are done first by formatting the query string as here and then using `simple_query`.

File: compute_tools/src/http/api.rs (hunk 9 above)
Comment (ololobus) [reply compute_tools/src/http/api.rs:103]: > Isn't this was Anastasia's work was all about? We just move her code into the GET side of the endpoint.

Yeah, this https://github.com/neondatabase/neon/pull/8917 added an `/installed_extensions` API

Don't really have a strong opinion. Converging to
```
POST /extensions
```
would work for me

Leaving `/installed_extensions` as also probably make sense because just 'extensions' is vague, and installed extensions != available

Comment (tristan957) [reply compute_tools/src/http/api.rs]: > Leaving /installed_extensions as also probably make sense because just 'extensions' is vague, and installed extensions != available

```
GET /extensions?installed=true
```

was my feedback in the original PR.

File: compute_tools/src/http/api.rs (hunk 3 above)
Comment (conradludgate) [reply compute_tools/src/http/api.rs]: turns out that it does - runtime inside a runtime problems.

File: compute_tools/src/compute.rs [hunk 11]
```diff
@@ -1367,6 +1368,43 @@ LIMIT 100",
...
+        if let Ok(installed_version) = version {
+            if installed_version == ext_version {
+                return Ok(installed_version);
+            }
+            let query = "ALTER EXTENSION $1 UPDATE TO $2";
+            db_client
+                .execute(query, &[&ext_name, &ext_version])
+                .context(format!("Failed to execute query: {}", query))?;
+        } else {
+            let query = "CREATE EXTENSION IF NOT EXISTS $1 WITH VERSION $2";
```
Comment (conradludgate) [inline compute_tools/src/compute.rs:1399]: `install_extension failed: Failed to execute query: CREATE EXTENSION IF NOT EXISTS $1 WITH VERSION $2`

Seems there's the same problem as in the other PR - this doesn't work with parameters.

Comment (conradludgate) [reply compute_tools/src/compute.rs:1399]: A second problem I found is that the version string needs to be quoted

**Comment by devjv** [comment]
@tristan957 this and the [other one](https://github.com/neondatabase/neon/pull/9395) should be fine to review

File: compute_tools/src/compute.rs [hunk 12]
```diff
@@ -1367,6 +1367,56 @@ LIMIT 100",
...
+        } else {
+            let query =
+                format!("CREATE EXTENSION IF NOT EXISTS {ext_name} WITH VERSION {ext_version}");
+            db_client
+                .simple_query(&query)
+                .await
+                .context(format!("Failed to execute query: {}", query))?;
+        }
+
+        Ok(ext_version.to_string())
```
Comment (tristan957) [inline compute_tools/src/compute.rs:1417]: I wonder if a more descriptive type would be useful rather than a string?

```c
typedef ExtVersion char *
```

This kind of thing from C

File: compute_tools/src/http/api.rs [hunk 13]
```diff
@@ -98,6 +100,38 @@ async fn routes(req: Request<Body>, compute: &Arc<ComputeNode>) -> Response<Body
             }
         }
 
+        (&Method::POST, "/extensions") => {
+            info!("serving /extensions POST request");
```
Comment (tristan957) [inline compute_tools/src/http/api.rs:104]: I think we are moving to make these `debug!`

File: compute_tools/src/compute.rs [hunk 14]
```diff
@@ -1367,6 +1367,56 @@ LIMIT 100",
...
+        &self,
+        ext_name: &str,
+        db_name: &str,
+        ext_version: &str,
+    ) -> Result<String> {
+        use tokio_postgres::config::Config;
+        use tokio_postgres::NoTls;
+
+        let mut conf = Config::from_str(self.connstr.as_str()).unwrap();
+        conf.dbname(db_name);
```
Comment (tristan957) [inline compute_tools/src/compute.rs:1380]: This is a good pattern. Didn't know about these APIs.

File: compute_tools/src/http/openapi_spec.yaml [hunk 15]
```diff
@@ -144,10 +144,38 @@ paths:
...
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/GenericError"
+
   /configure:
     post:
       tags:
-      - Configure
+        - Configure
```
Comment (tristan957) [inline compute_tools/src/http/openapi_spec.yaml:178]: Don't love the unrelated changes in this PR

Comment (tristan957) [reply compute_tools/src/http/openapi_spec.yaml:178]: I think a separate PR to fix formatting would be a better solution.

**Review by tristan957 — CHANGES_REQUESTED** [review]
I like the way this turned out.

**Review by ololobus — COMMENTED** [review]
Left similar comments as for the https://github.com/neondatabase/neon/pull/9395#pullrequestreview-2372863421

Otherwise, LGTM and almost ready to merge

File: compute_tools/src/http/api.rs (hunk 13 above)
Comment (tristan957) [reply compute_tools/src/http/api.rs:104]: I was going off @lassizci's comment here: https://github.com/neondatabase/neon/pull/9375#discussion_r1800925713. But perhaps the context is just completely different.

Comment (ololobus) [reply compute_tools/src/http/api.rs:104]: Yes, for metrics it's kinda all computes every dozen of seconds or so. Here, it is a small subset that will use Neon authorize

I think I was still a bit wrong saying that it's only once per DB. It will install extension once per DB, yes, but I guess local proxy doesn't know that it's installed, so will it call it every time we have a new connection?

Probably, Jere or Conrad can elaborate how often it will be called. Considering that only a few users will use it at start, I'd still probably keep the info for the ease of debugging, but do not have a strong opinion

File: compute_tools/src/http/openapi_spec.yaml (hunk 15 above)
Comment (devjv) [reply compute_tools/src/http/openapi_spec.yaml:178]: Zed is quite trigger happy to format stuff, I will remove these/ 🙇

File: compute_tools/src/http/api.rs (hunk 13 above)
Comment (devjv) [reply compute_tools/src/http/api.rs:104]: > install the extension once per database per boot (then caching that it's installed)
This is how it should work.

We discussed this with @lassizci that we can always drop this to `debug` if this ends up being too noisy.

**Comment by devjv** [comment]
Added test and addressed other feedback as well

File: compute_tools/src/http/openapi_spec.yaml [hunk 16]
```diff
@@ -144,6 +144,34 @@ paths:
...
+            schema:
+              $ref: "#/components/schemas/ExtensionInstallRequest"
+      responses:
+        200:
+          description: Result from extension installation
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/ExtensionInstallResult"
+        500:
```
Comment (ololobus) [inline compute_tools/src/http/openapi_spec.yaml:168]: NIT

```suggestion
        412:
          description: |
            Compute is in the wrong state for processing the request.
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/GenericError"
        500:
```

File: compute_tools/src/compute.rs [hunk 17]
```diff
@@ -1367,6 +1368,56 @@ LIMIT 100",
...
+
+        let (db_client, conn) = conf
+            .connect(NoTls)
+            .await
+            .context("Failed to connect to the database")?;
+        tokio::spawn(conn);
+
+        let version_query = "SELECT extversion FROM pg_extension WHERE extname = $1";
+        let version: Option<String> = db_client
```
Comment (ololobus) [inline compute_tools/src/compute.rs:1390]: NIT
```suggestion
        let version: Option<ExtVersion> = db_client
```
?

**Review by ololobus — APPROVED** [review]
